path_dynamic_roles.go

```diff
@@ -0,0 +1,317 @@
+package openldap
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"path"
+	"time"
+
+	"github.com/go-ldap/ldif"
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/vault/sdk/framework"
+	"github.com/hashicorp/vault/sdk/helper/parseutil"
+	"github.com/hashicorp/vault/sdk/logical"
+	"github.com/mitchellh/mapstructure"
+)
+
+const (
+	secretCredsType = "creds"
+
+	dynamicRolePath = "role/"
+	dynamicCredPath = "cred/"
+)
+
+func (b *backend) pathDynamicRoles() []*framework.Path {
+	return []*framework.Path{
+		// POST/GET/DELETE role/:name
+		{
+			Pattern: path.Join(dynamicRolePath, framework.GenericNameRegex("name")),
+			Fields: map[string]*framework.FieldSchema{
+				"name": {
+					Type:        framework.TypeLowerCaseString,
+					Description: "Name of the role (lowercase)",
+					Required:    true,
+				},
+				"creation_ldif": {
+					Type:        framework.TypeString,
+					Description: "LDIF string used to create new entities within OpenLDAP. This LDIF can be templated.",
+					Required:    true,
+				},
+				"deletion_ldif": {
+					Type:        framework.TypeString,
+					Description: "LDIF string used to delete entities created within OpenLDAP. This LDIF can be templated.",
+					Required:    true,
+				},
+				"rollback_ldif": {
+					Type:        framework.TypeString,
+					Description: "LDIF string used to rollback changes in the event of a failure to create credentials. This LDIF can be templated.",
+				},
+				"username_template": {
+					Type:        framework.TypeString,
+					Description: "The template used to create a username",
+				},
+				"default_ttl": {
+					Type:        framework.TypeDurationSecond,
+					Description: "Default TTL for dynamic credentials",
+				},
+				"max_ttl": {
+					Type:        framework.TypeDurationSecond,
+					Description: "Max TTL a dynamic credential can be extended to",
+				},
+			},
+			ExistenceCheck: b.pathDynamicRoleExistenceCheck,
```

Comment:
Wondering if the existence check is needed here? It looks like a write is processed the same in `pathDynamicRoleCreateUpdate()` whether the role exists or not.

Reply:
It is needed, but not for the create/update endpoint. It was needed for one of the other non-endpoints. I don't recall exactly which one off hand (renewal? revocation?). I remember it was giving me issues if I didn't specify it.